src/server/controllers/oktaIdxShared.ts

```diff
@@ -0,0 +1,257 @@
+		stateHandleExpiresAt: challengeEmailResponse.expiresAt,
+		userState: 'ACTIVE_PASSWORD_ONLY',
+	});
+};
+
+/**
+ * @name forceUserIntoActiveState
+ * @description IMPORTANT: Use this method to force a user into the ACTIVE state from a non-ACTIVE state. Please read the following:
+ *
+ * In order for a user in a non-ACTIVE state to be able to receive a passcode, they must be in the ACTIVE state, so we have to
+ * force them into the ACTIVE state.
+ *
+ * The best way to do this is to first deactivate the user, which works on all user states and puts them
+ * into the DEPROVISIONED state.
+ * Then we can activate the user, which will put them into the PROVISIONED state and return us a recovery token.
+ * We then use the recovery token to set a placeholder password for the user, which transitions them into the ACTIVE state,
+ * and also makes sure to unset the `emailValidated` and `passwordSetSecurely` flags if they are currently set to `true`.
+ *
+ * This will put the user in one of the following ACTIVE states:
+ * 1. ACTIVE users - has email + password authenticator (okta idx email verified)
+ * 2. ACTIVE users - has only password authenticator (okta idx email not verified)
+ *
+ * They can't be in the 3. ACTIVE users - has only email authenticator (SOCIAL users, no password), as these users will
+ * have a password.
+ *
+ * From this point the calling function will be able to send the user a passcode.
+ */
+export const forceUserIntoActiveState = async ({
+	req,
+	user,
+}: {
+	req: Request;
+	user: UserResponse;
+}) => {
+	// 1. deactivate the user
+	try {
+		await deactivateUser({
+			id: user.id,
+			ip: req.ip,
+		});
+		trackMetric('OktaDeactivateUser::Success');
+	} catch (error) {
+		trackMetric('OktaDeactivateUser::Failure');
+		logger.error(
+			'Okta user deactivation failed',
+			error instanceof OktaError ? error.message : error,
+		);
+		throw error;
+	}
+
+	// 2. activate the user
+	try {
+		const tokenResponse = await activateUser({
+			id: user.id,
+			ip: req.ip,
+		});
+		if (!tokenResponse?.token.length) {
+			throw new OktaError({
+				message: `Okta force activation failed: missing activation token`,
+			});
+		}
+
+		// 3. use the recovery token to set a placeholder password for the user
+		// Validate the token
+		const { stateToken } = await validateRecoveryToken({
+			recoveryToken: tokenResponse.token,
+			ip: req.ip,
+		});
+		// Check if state token is defined
+		if (!stateToken) {
+			throw new OktaError({
+				message:
+					'Okta set placeholder password failed: state token is undefined',
+			});
+		}
+		// Set the placeholder password as a cryptographically secure UUID
+		const placeholderPassword = crypto.randomUUID();
+		await resetPassword(
+			{
+				stateToken,
+				newPassword: placeholderPassword,
+			},
+			req.ip,
+		);
+
+		// Unset the emailValidated and passwordSetSecurely flags
+		await validateEmailAndPasswordSetSecurely({
+			id: user.id,
+			ip: req.ip,
+			flagStatus: false,
+		});
```

Comment:
Might be missing something, but this seems very similar to what [dangerouslySetPlaceholderPassword](https://github.com/guardian/gateway/blob/b9e2e18eb61217c41da2e7ea485173a89597a727/src/server/lib/okta/dangerouslySetPlaceholderPassword.ts#L36C16-L36C49) is doing?

Reply:
that's a good spot! I'll refactor that.